iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c

```diff
@@ -0,0 +1,441 @@
+
+        bool maxTempUpdated = false;
+        UpdateTemperatureAndStatistics(targetTemperature, &maxTempUpdated);
+
+        // The device needs to let the service know that it has received the targetTemperature desired property.
+        SendTargetTemperatureReport(deviceClient, targetTemperature, g_statusSuccess, version, g_temperaturePropertyResponseDescription);
+
+        if (maxTempUpdated)
+        {
+            // If the Maximum temperature has been updated, we also report this as a property.
+            SendMaxTemperatureSinceReboot(deviceClient);
+        }
+    }
+
+    json_value_free(rootValue);
+    free(jsonStr);
+}
+
+//
+// Thermostat_SendCurrentTemperature sends a PnP telemetry indicating the current temperature
+//
+void Thermostat_SendCurrentTemperature(IOTHUB_DEVICE_CLIENT_HANDLE deviceClient) 
+{
+    IOTHUB_MESSAGE_HANDLE messageHandle = NULL;
+    IOTHUB_CLIENT_RESULT iothubResult;
+
+    char temperatureStringBuffer[32];
+
+    if (snprintf(temperatureStringBuffer, sizeof(temperatureStringBuffer), g_TemperatureTelemetryBodyFormat, g_currentTemperature) < 0)
+    {
+        LogError("snprintf of current temperature telemetry failed");
+    }
+    else if ((messageHandle = IoTHubMessage_CreateFromString(temperatureStringBuffer)) == NULL)
+    {
+        LogError("IoTHubMessage_CreateFromString failed");
+    }
+    else if ((iothubResult = IoTHubDeviceClient_SendEventAsync(deviceClient, messageHandle, NULL, NULL)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to send telemetry message, error=%d", iothubResult);
+    }
+
+    IoTHubMessage_Destroy(messageHandle);
+}
+
+int main(void)
+{
+    IOTHUB_DEVICE_CLIENT_HANDLE deviceClient = NULL;
+    const char* connectionString;
+
+    if ((connectionString = getenv(g_ConnectionStringEnvironmentVariable)) == NULL)
+    {
+        LogError("Cannot read environment variable %s", g_ConnectionStringEnvironmentVariable);
+    }
+    else if ((deviceClient = PnPHelper_CreateDeviceClientHandle(connectionString, g_ModelId, g_hubClientTraceEnabled, Thermostat_DeviceMethodCallback, Thermostat_DeviceTwinCallback)) == NULL)
+    {
+        LogError("Failure creating Iothub device");
+    }
+    else
+    {
+        LogInfo("Successfully created device client handle.  Hit Control-C to exit program\n");
+
+        SendMaxTemperatureSinceReboot(deviceClient);
+
+        while (true)
+        {
+            // Wake up periodically to send telemetry.
+            // IOTHUB_DEVICE_CLIENT_HANDLE brings in the IoTHub device convenience layer, which means that the IoTHub SDK itself 
+            // spins a worker thread to perform all operations.
+            Thermostat_SendCurrentTemperature(deviceClient);
+            ThreadAPI_Sleep(g_sleepBetweenTelemetrySends);
+        }
+
+        // Clean up the iothub sdk handle
+        IoTHubDeviceClient_Destroy(deviceClient);
+        // Free all the sdk subsystem
+        IoTHub_Deinit();        
```

Comment:
should we have delete api corresponding to create?

Reply:
Very good question.  If this were a true API, then yes we'd need an analog for delete.  The issue here is rest of the sample has direct access to the underlying deviceClient handle such as sending twins/properties/etc.  So for destruction I'm inclined to follow similar rule.  The creation helper is about convenience, not ownership.